sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -0,0 +1,218 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: profile=True
+#
+# Author: Andreas Mueller
+#
+# Licence: BSD 3 clause
+
+import numpy as np
+cimport numpy as np
+cimport cython
+
+from libc.math cimport sqrt
+
+from ..metrics import euclidean_distances
+from ._k_means import _centers_dense
+from ..utils.fixes import partition
+
+
+cdef double euclidian_dist(double* a, double* b, int n_features) nogil:
+    cdef double result, tmp
+    result = 0
+    cdef int i
+    for i in range(n_features):
+        tmp = (a[i] - b[i])
+        result += tmp * tmp
+    return sqrt(result)
+
+
+cdef update_labels_distances_inplace(
+        double* X, double* centers, double[:, :] center_distances, int[:] labels,
+        double[:, :] lower_bounds, double[:] upper_bounds, int n_samples, int
+        n_features, int n_clusters):
+    """
+    Calculate upper and lower bounds for each sample.
+
+    Given X, centers and the pairwise distances divided by 2.0 between the
+    centers this calculates the upper bounds and lower bounds for each sample.
+    The upper bound for each sample is set to the distance between the sample
+    and the closest center.
+
+    The lower bound for each sample is a one-dimensional array of n_clusters.
+    For each sample i assume that the previously assigned cluster is c1 and the
+    previous closest distance is dist, for a new cluster c2, the
+    lower_bound[i][c2] is set to distance between the sample and this new
+    cluster, if and only if dist > center_distances[c1][c2]. This prevents
+    computation of unnecessary distances for each sample to the clusters that
+    it is unlikely to be assigned to.
+    """
```

Comment:
The docstring is difficult to understand. Could you modify yout docstring to be more like that : 
```
""" Calculate upper and lower bounds for each sample.

Parameters
----------
X_ : nd-array, shape (n_samples, n_features)

centers : nd-array, ...

center_distances : nd-array, ...
   Pairwise half distances.
...
lower_bounds : nd-array, ...
    For a sample i...

upper_bounds : nd-array, ...
    The upper bound is defined for each sample as the distance between the sample and the closest center.

...
```